AIM/sourcefiles/AIMV1-0-3.cpp

```diff
@@ -1,496 +1,498 @@
+	}
+
+	int ix_builder(int *inrange_reslist, int inrange_reslist_len, int *residues_start, int *residues_fin, int *inrange_ix) {
+		int counter = 0;
+		for (int index = 0; index < inrange_reslist_len; index++) {
+			int resnum = inrange_reslist[index];
+			int start = residues_start[resnum];
+			int fin = residues_fin[resnum];
+			for (int atnum = start; atnum < (fin + 1); atnum++) {
+				inrange_ix[counter] = atnum;
+				counter++;
+			}
+		}
+		return counter;
+	}
+
+	void PrepTCC(int *AllAmGroups, int *PrePro, float *TCC_v, int *relevant_groups, int relgrouplen, float *vPro, float *vGen, float alphaPro, float alphaGen, float *WS_positions, float *halfbox, float *boxdims) {
+		int C, O, N, AmGroup;
+		float CO[3], CN[3], z[3];
+        for (int count = 0; count < relgrouplen; count++) {
+            AmGroup = relevant_groups[count];
+			C = AllAmGroups[AmGroup * 18 + 6];
+			O = AllAmGroups[AmGroup * 18 + 7];
+			N = AllAmGroups[AmGroup * 18 + 9];
+			PBC_diff(&WS_positions[O * 3], &WS_positions[C * 3], halfbox, boxdims, CO);
+			PBC_diff(&WS_positions[N * 3], &WS_positions[C * 3], halfbox, boxdims, CN);
+			vecnorm(CO);
+			project(CO, CN);
+			vecnorm(CN);
+			crossprod(CO, CN, z);
+
+			if (PrePro[AmGroup] == 0) {
+				for (int a = 0; a < 6; a++) {
+					for (int c = 0; c < 3; c++) {
+						TCC_v[AmGroup * 18 + a * 3 + c] = (CO[c] * vGen[a * 3] + CN[c] * vGen[a * 3 + 1] + z[c] * vGen[a * 3 + 2])*alphaGen;
+					}
+				}
+			}
+			else {
+				for (int a = 0; a < 6; a++) {
+					for (int c = 0; c < 3; c++) {
+						TCC_v[AmGroup * 18 + a * 3 + c] = (CO[c] * vPro[a * 3] + CN[c] * vPro[a * 3 + 1] + z[c] * vPro[a * 3 + 2])*alphaPro;
+					}
+				}
+			}
+		}
+
+	}
+
+	float CalcTCC(int AmGroupi, int AmGroupj, int *AllAmGroups, int *PrePro, float *TCC_v, float *qGen, float *qPro, float *dqGen, float *dqPro, float TCC_4PiEps, float *WS_positions, float *halfbox, float *boxdims) {
+		float J = 0;
+		int atom_a, atom_b;
+		float va[3], vb[3], d[3];
+		float r2, ir2, ir, ir3, ir5;
+		float qi, qj[6], dqi, dqj[6];
+		int tyi, tyj;
+
+		tyi = PrePro[AmGroupi];
+		tyj = PrePro[AmGroupj];
+
+		for (int b = 0; b < 6; b++) {
+			qj[b] = (tyj * qPro[b] + (1-tyj) * qGen[b]);
+			dqj[b] = (tyj * dqPro[b] + (1-tyj) * dqGen[b]);
+		}
+
+		for (int a = 0; a < 6; a++) {
+			qi = (tyi * qPro[a] + (1-tyi) * qGen[a]);
+			dqi = (tyi * dqPro[a] + (1-tyi) * dqGen[a]);
+			atom_a = AllAmGroups[AmGroupi * 18 + 6 + a];
+			for (int c = 0; c < 3; c++) {
+				va[c] = TCC_v[AmGroupi * 18 + a * 3 + c];
+			}
+			for (int b = 0; b < 6; b++) {
+				atom_b = AllAmGroups[AmGroupj * 18 + 6 + b];
+				PBC_diff(&WS_positions[atom_b * 3], &WS_positions[atom_a * 3], halfbox, boxdims, d);
+				r2 = dotprod(d, d);
+				if (r2 < 0.01) {r2 = 1.00;}
+				ir2 = 1 / r2;
+				ir = sqrt(ir2);
+				ir3 = ir * ir2;
+				ir5 = ir3 * ir2;
+
+				for (int c = 0; c < 3; c++) {
+					vb[c] = TCC_v[AmGroupj * 18 + b * 3 + c];
+				}
+
+				J -= 3 * ir5 * qi * qj[b] * dotprod(vb, d) * dotprod(va, d);
```

Comment:
I would also add comment with reference and potential note on the sign - vector difference from Eq A6 of the 2006 paper. Also note that the sign used in the original paper for the qd-parameters was wrong (see erratum) therefore the signs here are only correct if you use the opposite sign of the parameters for dq as to what is given in the paper....

I looked up [AIM](https://github.com/Kimvana/AIM/tree/Version-1.0-(main))/[AIM](https://github.com/Kimvana/AIM/tree/Version-1.0-(main)/AIM)/[sourcefiles](https://github.com/Kimvana/AIM/tree/Version-1.0-(main)/AIM/sourcefiles)
/Map-C_TCC.dat
There the sign of the dq parameters are as in the paper, so the original direction given here was actually the correct one according to the erratum.